File: spec/ce2/relations/set_spec.rb
require File.expand_path("#{File.dirname(__FILE__)}/../ce2_spec_helper")

module Relations
  describe Set do

    attr_reader :set
    before do
      @set = Answer.set
    end

    describe "#initialize" do
      it "automatically has a string-valued :id attribute" do
        set.attributes_by_name[:id].type.should == :string
      end
    end

    describe "#define_attribute" do
      it "adds an Attribute with the given name and type and self as its #set to the #attributes_by_name hash" do
        attribute = set.attributes_by_name[:body]
        attribute.name.should == :body
        attribute.type.should == :string
      end
    end

    describe "#attributes" do
      it "returns the #values of #attributes_by_name" do
        set.attributes.should == set.attributes_by_name.values
      end
    end

    describe "#insert" do
      it "calls Origin.insert with the Set's #global_name and #field_values_by_attribute_name" do
        tuple = Answer.new(:body => "Brown Rice", :correct => true)
        mock(Origin).insert(set.global_name, tuple.field_values_by_attribute_name)
        set.insert(tuple)
      end
    end

    describe "#create" do
      it "instantiates an instance of #tuple_class with the given attributes, #inserts it, and returns it" do
        mock(set).insert(anything) do |tuple|
          tuple.class.should == set.tuple_class
          tuple.body.should == "Brown Rice"
          tuple.correct.should == true
        end

        tuple = set.create(:body => "Brown Rice", :correct => true)
        tuple.body.should == "Brown Rice"
      end
    end

    describe "#tuples" do
      it "executes a select all SQL query against the database and returns Tuples corresponding to its result" do
        tuple_1_id = set.create(:body => "Quinoa", :correct => true).id
        tuple_2_id = set.create(:body => "White Rice", :correct => false).id
        tuple_3_id = set.create(:body => "Pearled Barley", :correct => false).id

        mock.proxy(Origin).read(set.tuple_class, "select answers.id, answers.body, answers.question_id, answers.correct from answers;")

        tuples = set.tuples

        retrieved_tuple_1 = tuples.find {|t| t.id == tuple_1_id }
        retrieved_tuple_1.body.should == "Quinoa"
        retrieved_tuple_1.correct.should == true

        retrieved_tuple_2 = tuples.find {|t| t.id == tuple_2_id }
        retrieved_tuple_2.body.should == "White Rice"
        retrieved_tuple_2.correct.should == false

        retrieved_tuple_3 = tuples.find {|t| t.id == tuple_3_id }
        retrieved_tuple_3.body.should == "Pearled Barley"
        retrieved_tuple_3.correct.should == false
      end
    end

    describe "#to_sql" do
      it "returns a select statement for only the columns declared as Attributes on the Set" do
        columns = set.attributes.map {|a| a.to_sql }.join(", ")
        set.to_sql.should == "select #{columns} from #{set.global_name};"
      end
    end
  end
end

File: spec/javascript/lib/model/relations/set_spec.js
//= require "../../../hyperarchy_spec_helper"

Screw.Unit(function(c) { with(c) {
  describe("Model.Relations.Set", function() {
    var set;
    before(function() {
      set = new Model.Relations.Set("programming_languages");
    });


    describe("#define_attribute", function() {
      var attribute;
      before(function() {
        attribute = set.define_attribute("family_id", "string");
      });

      it("adds an Attribute with the given name and type to #attributes_by_name and returns it", function() {
        expect(attribute).to(equal, set.attributes_by_name.family_id);
        expect(attribute.constructor).to(equal, Model.Attribute);
        expect(attribute.name).to(equal, 'family_id');
        expect(attribute.type).to(equal, 'string');
      });
    });

    describe("#insert", function() {
      it("adds the given Tuple to the array returned by #all", function() {
        var mock_tuple = {};

        expect(set.all()).to_not(contain, mock_tuple);
        set.insert(mock_tuple)
        expect(set.all()).to(contain, mock_tuple);
      });
    });
  });
}});